presto-main/src/test/java/com/facebook/presto/sql/TestExpressionInterpreter.java

```diff
@@ -577,6 +577,7 @@ public void testCastToString()
 
         // string
         assertOptimizedEquals("cast('xyz' as VARCHAR)", "'xyz'");
+        assertOptimizedEquals("cast(cast('abcxyz' as VARCHAR(3)) as VARCHAR(5))", "'abc'");
```

Comment:
Maybe also add tests where this will not be optimized. Though you might have to check plan or something to verify.

Reply:
It is actually hard to test here. If the value is constant, even if this will not be optimized away here. It will likely be evaluated by invoking the `methodHandle`.  It only will be a problem when there is a massive number of  function arguments (for example, `RowConstructor`) contain such non optimized casts.